Mathlib/Data/List/Shortlex.lean

```diff
@@ -0,0 +1,296 @@
+      rw [← hp.1] at h
+      exact h
+    | right a h =>
+      right
+      rw [Prod.mk.injEq] at hp
+      rw [← hp.2] at h
+      exact ⟨hp.1, h⟩
+  intro hpq
+  rcases hpq with h1 | h2
+  · exact shorter_than h1
+  exact len_equal h2.1 h2.2
+
+theorem cons_iff [IsIrrefl α r] {a : α} {s t : List α} : Shortlex r (a :: s) (a :: t) ↔
+    Shortlex r s t := by
+  constructor
+  · intro h
+    rcases shortlex_def.mp h with h1 | h2
+    · simp only [List.length_cons, Nat.succ_eq_add_one, add_lt_add_iff_right] at h1
+      exact shorter_than h1
+    simp only [List.length_cons, Nat.succ_eq_add_one, add_left_inj] at h2
+    rw [List.Lex.cons_iff] at h2
+    exact len_equal h2.1 h2.2
+  intro h
+  rcases shortlex_def.mp h with h1 | h2
+  · apply shorter_than
+    simp only [List.length_cons, Nat.succ_eq_add_one, add_lt_add_iff_right]
+    exact h1
+  apply len_equal
+  · simp only [List.length_cons, Nat.succ_eq_add_one, add_left_inj]
+    exact h2.1
+  rw [List.Lex.cons_iff]
+  exact h2.2
+
+@[simp]
+theorem not_nil_right {s : List α} : ¬ Shortlex r s [] := by
+  intro h
+  rcases shortlex_def.mp h with h1 | h2
+  · simp only [List.length_nil, not_lt_zero'] at h1
+  exact List.Lex.not_nil_right _ _ h2.2
+
+theorem nil_left_or_eq_nil (s : List α) : Shortlex r [] s ∨ s = [] := by
+  cases s with
+  | nil => right; rfl
+  | cons head tail =>
+    left
+    apply shorter_than
+    simp only [List.length_nil, List.length_cons, Nat.succ_eq_add_one, lt_add_iff_pos_left,
+      add_pos_iff, zero_lt_one, or_true]
+
+@[simp]
+theorem singleton_iff (a b : α) : Shortlex r [a] [b] ↔ r a b := by
+  constructor
+  · intro h
+    rcases shortlex_def.mp h with h1 | h2
+    · simp only [List.length_singleton, lt_self_iff_false] at h1
+    exact (List.Lex.singleton_iff _ _).mp h2.2
+  exact fun h =>  len_equal rfl ((List.Lex.singleton_iff _ _).mpr h)
+
+instance isOrderConnected [IsOrderConnected α r] [IsTrichotomous α r] :
+    IsOrderConnected (List α) (Shortlex r) where
+  conn := by
+    intro a b c ac
+    rcases shortlex_def.mp ac with h1 | h2
+    · rcases Nat.lt_or_ge a.length b.length with h3 | h4
+      · left; exact shorter_than h3
+      rcases Nat.lt_or_ge b.length c.length with h3 | h4
+      · right; exact shorter_than h3
+      omega
+    rcases Nat.lt_or_ge a.length b.length with h3 | h4
+    · left; exact shorter_than h3
+    rcases Nat.lt_or_ge b.length c.length with h3 | h4
+    · right; exact shorter_than h3
+    have hab : a.length = b.length := by omega
+    have hbc : b.length = c.length := by omega
+    rcases List.Lex.isOrderConnected.aux r a b c h2.2 with h5 | h6
+    · left
+      exact len_equal hab h5
+    right
+    exact len_equal hbc h6
+
+instance isTrichotomous [IsTrichotomous α r] : IsTrichotomous (List α) (Shortlex r) where
+  trichotomous := by
+    intro a b
+    have : a.length < b.length ∨ a.length = b.length ∨ a.length > b.length :=
+      trichotomous a.length b.length
+    rcases this with h1 | h23
```

Comment:
Just nitpicking here but you could write this as `rcases this with h1 | h2 | h3` which saves you the trouble of having to do another `rcases` two lines below. Same goes for line 165 with the extra `rcases` on 171.

Reply:
oh nice, thanks!